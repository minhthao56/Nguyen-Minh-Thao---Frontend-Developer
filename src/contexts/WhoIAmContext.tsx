import { createContext, useContext } from "react";
import { AuthResp } from "../services/user";

export interface WhoIAmContextReturn extends AuthResp {
  isLoading?: boolean;
}

const WhoIAmContextContext = createContext<WhoIAmContextReturn>({
  email: "",
  token: "",
  isLoading: false,
});
WhoIAmContextContext.displayName = "WhoIAmContextContext";
const useWhoIAmContext = () => useContext(WhoIAmContextContext);
export { WhoIAmContextContext, useWhoIAmContext };
