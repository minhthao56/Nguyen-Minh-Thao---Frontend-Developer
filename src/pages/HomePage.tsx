import Banner from "../components/Banner";
import FormFilter from "../components/FormFilter";
import RocketCard from "../components/RocketCard";
import { useQuery } from "@tanstack/react-query";
import { apiRocket } from "../services/rocket";
import Loading from "../components/Loading";
import { useState } from "react";
import debounce from "lodash/debounce";
import { useWhoIAmContext } from "../contexts/WhoIAmContext";
import { Link } from "react-router-dom";

export default function HomePage() {
  const [search, setSearch] = useState("");
  const { email } = useWhoIAmContext();

  const { data, isLoading, isError, error, refetch } = useQuery({
    queryKey: ["rockets", search],
    queryFn: async () => {
      return await apiRocket.getRockets(search);
    },
    enabled: Boolean(localStorage.getItem("token")),
  });

  const onChange = debounce((e: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value);
  }, 500);

  return (
    <>
      <Banner />
      {email ? (
        <div className="max-w-screen-2xl m-auto">
          <FormFilter
            onChange={onChange}
            onSearch={() => {
              refetch();
            }}
          />
          <h1 className="text-2xl uppercase  m-4">List rockets</h1>
          {isError && (
            <p className="text-red-800 text-center mb-4">
              {JSON.stringify(error instanceof Error ? error.message : error)}
            </p>
          )}
          {isLoading ? (
            <Loading />
          ) : (
            <div className="grid gap-4 p-4 xl:grid-cols-4 lg:grid-cols-3 md:grid-cols-2 sm:grid-cols-1 place-items-center">
              {data?.map((rocket, i) => {
                return (
                  <RocketCard
                    image={rocket?.flickr_images?.[0]}
                    description={rocket.description}
                    name={rocket.name}
                    country={rocket.country}
                    key={i}
                  />
                );
              })}
            </div>
          )}
        </div>
      ) : (
        <p className="w-full text-center text-3xl py-24">
          Please{" "}
          <Link to={"/login"} className="text-blue-500 underline">
            Login
          </Link>{" "}
          or{" "}
          <Link to={"/register"} className="text-blue-600 underline">
            Sign In
          </Link>{" "}
          to see list rocket
        </p>
      )}
    </>
  );
}
