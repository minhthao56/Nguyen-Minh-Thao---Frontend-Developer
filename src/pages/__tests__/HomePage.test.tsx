import { render, screen, waitFor } from "@testing-library/react";
import HomePage from "../HomePage";
import { useQuery } from "@tanstack/react-query";
import userEvent from "@testing-library/user-event";
import mockData from "../../services/rocket.json";
import { useWhoIAmContext } from "../../contexts/WhoIAmContext";
import { BrowserRouter } from "react-router-dom";

jest.mock("@tanstack/react-query", () => {
  return {
    ...jest.requireActual("@tanstack/react-query"),
    useMutation: jest.fn(),
    useQuery: jest.fn(),
  };
});

jest.mock("../../contexts/WhoIAmContext", () => {
  return {
    useWhoIAmContext: jest.fn(),
  };
});

jest.mock("../../services/rocket", () => {
  return {
    apiRocket: {
      getRockets: jest.fn(),
    },
  };
});

describe("<HomePage/>", () => {
  const refetch = jest.fn();

  beforeEach(() => {
    (useQuery as jest.Mock).mockReturnValue({
      data: mockData,
      refetch,
    });
    (useWhoIAmContext as jest.Mock).mockReturnValue({ email: "123@gmail.com" });
  });

  const renderComponent = () => {
    return render(
      <BrowserRouter>
        <HomePage />
      </BrowserRouter>
    );
  };

  it("should match to snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it("should call refetch function", () => {
    renderComponent();

    userEvent.click(screen.getByRole("button", { name: "Search" }));

    expect(refetch).toBeCalled();
  });

  it("should call useQuery hook when typing search", async () => {
    renderComponent();

    userEvent.type(screen.getByPlaceholderText("Search By Name"), "test");
    await waitFor(() => {
      expect(useQuery).toBeCalled();
    });
  });

  it("should show error message", () => {
    (useQuery as jest.Mock).mockReturnValue({
      isError: true,
      error: new Error("Message error"),
    });

    renderComponent();

    expect(
      screen.getByText(JSON.stringify("Message error"))
    ).toBeInTheDocument();
  });

  it("should show loading", () => {
    (useQuery as jest.Mock).mockReturnValue({
      isLoading: true,
    });

    renderComponent();

    expect(screen.getByTestId("Loading")).toBeInTheDocument();
  });

  it("should not display list rocket when non-auth", () => {
    (useWhoIAmContext as jest.Mock).mockReturnValue({ email: "" });

    const { container } = renderComponent();

    expect(container).toMatchSnapshot();

    expect(screen.getByText(/to see list rocket/)).toBeInTheDocument();
  });
});
