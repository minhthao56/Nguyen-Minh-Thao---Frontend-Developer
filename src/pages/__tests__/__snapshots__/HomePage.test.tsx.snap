// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage/> should match to snapshot 1`] = `
<div>
  <div
    class="bg-banner bg-cover bg-no-repeat h-96 items-center flex flex-col justify-center"
  >
    <h1
      class="text-white font-bold text-4xl"
    >
      SHOP SPACEX
    </h1>
    <p
      class="text-white font-thin"
    >
      Lorem ipsum dolor sit amet consectetur, adipisicing elit.
    </p>
  </div>
  <div
    class="max-w-screen-2xl m-auto"
  >
    <div
      class="m-4"
    >
      <h1
        class="text-2xl uppercase mb-6"
      >
        Search SpaceX
      </h1>
      <div
        class="flex"
      >
        <input
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-black focus:black block w-full p-2.5 undefined"
          placeholder="Search By Name"
          type="text"
          value=""
        />
        <button
          class="text-white bg-black focus:ring-4 focus:outline-none focus:ring-black font-medium rounded-lg text-sm w-full px-5 py-2.5 text-center ml-5"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
    <h1
      class="text-2xl uppercase  m-4"
    >
      List rockets
    </h1>
    <div
      class="grid gap-4 p-4 xl:grid-cols-4 lg:grid-cols-3 md:grid-cols-2 sm:grid-cols-1 place-items-center"
    >
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <img
          alt="Sunset in the mountains"
          class="w-full  h-72"
          src="https://imgur.com/DaCfMsj.jpg"
        />
        <div
          class="px-6 py-4"
        >
          <div
            class="font-bold text-xl mb-2"
          >
            Falcon 1
          </div>
          <p
            class="text-gray-700 text-base line-clamp-2"
          >
            The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
          </p>
        </div>
        <div
          class="px-6 pb-2"
        >
          <span
            class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2"
          >
            Republic of the Marshall Islands
          </span>
        </div>
      </div>
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <img
          alt="Sunset in the mountains"
          class="w-full  h-72"
          src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
        />
        <div
          class="px-6 py-4"
        >
          <div
            class="font-bold text-xl mb-2"
          >
            Falcon 9
          </div>
          <p
            class="text-gray-700 text-base line-clamp-2"
          >
            Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
          </p>
        </div>
        <div
          class="px-6 pb-2"
        >
          <span
            class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2"
          >
            United States
          </span>
        </div>
      </div>
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <img
          alt="Sunset in the mountains"
          class="w-full  h-72"
          src="https://farm5.staticflickr.com/4599/38583829295_581f34dd84_b.jpg"
        />
        <div
          class="px-6 py-4"
        >
          <div
            class="font-bold text-xl mb-2"
          >
            Falcon Heavy
          </div>
          <p
            class="text-gray-700 text-base line-clamp-2"
          >
            With the ability to lift into orbit over 54 metric tons (119,000 lb)--a mass equivalent to a 737 jetliner loaded with passengers, crew, luggage and fuel--Falcon Heavy can lift more than twice the payload of the next closest operational vehicle, the Delta IV Heavy, at one-third the cost.
          </p>
        </div>
        <div
          class="px-6 pb-2"
        >
          <span
            class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2"
          >
            United States
          </span>
        </div>
      </div>
      <div
        class="max-w-sm rounded overflow-hidden shadow-lg"
      >
        <img
          alt="Sunset in the mountains"
          class="w-full  h-72"
          src="https://live.staticflickr.com/65535/48954138962_ee541e6755_b.jpg"
        />
        <div
          class="px-6 py-4"
        >
          <div
            class="font-bold text-xl mb-2"
          >
            Starship
          </div>
          <p
            class="text-gray-700 text-base line-clamp-2"
          >
            Starship and Super Heavy Rocket represent a fully reusable transportation system designed to service all Earth orbit needs as well as the Moon and Mars. This two-stage vehicle — composed of the Super Heavy rocket (booster) and Starship (ship) — will eventually replace Falcon 9, Falcon Heavy and Dragon.
          </p>
        </div>
        <div
          class="px-6 pb-2"
        >
          <span
            class="inline-block bg-gray-200 rounded-full px-3 py-1 text-sm font-semibold text-gray-700 mr-2 mb-2"
          >
            United States
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<HomePage/> should not display list rocket when non-auth 1`] = `
<div>
  <div
    class="bg-banner bg-cover bg-no-repeat h-96 items-center flex flex-col justify-center"
  >
    <h1
      class="text-white font-bold text-4xl"
    >
      SHOP SPACEX
    </h1>
    <p
      class="text-white font-thin"
    >
      Lorem ipsum dolor sit amet consectetur, adipisicing elit.
    </p>
  </div>
  <p
    class="w-full text-center text-3xl py-24"
  >
    Please
     
    <a
      class="text-blue-500 underline"
      href="/login"
    >
      Login
    </a>
     
    or
     
    <a
      class="text-blue-600 underline"
      href="/register"
    >
      Sign In
    </a>
     
    to see list rocket
  </p>
</div>
`;
